ocaml/xapi/xapi_vdi.ml

```diff
@@ -861,6 +861,35 @@ let export_changed_blocks ~__context ~vdi_from ~vdi_to =
   Storage_access.transform_storage_exn
     (fun () ->
        C.VDI.export_changed_blocks ~dbg:(Ref.string_of task) ~sr ~vdi_from ~vdi_to
-    );
+    )
+
+let get_nbd_info ~__context ~self =
+  if (Db.VDI.get_type ~__context ~self) = `cbt_metadata then begin
+    error "VDI.get_nbd_info: called with a VDI of type cbt_metadata (at %s)" __LOC__;
+    raise (Api_errors.Server_error (Api_errors.vdi_incompatible_type, [ Ref.string_of self; Record_util.vdi_type_to_string `cbt_metadata ]))
+  end;
+
+  let sr = Db.VDI.get_SR ~__context ~self in
+  let hosts_with_attached_pbds =
+    Db.SR.get_PBDs ~__context ~self:sr
+    |> List.filter (fun pbd -> Db.PBD.get_currently_attached ~__context ~self:pbd)
+    |> List.map (fun pbd -> Db.PBD.get_host ~__context ~self:pbd)
+  in
+  let get_ips host =
+    let get_ips pif =
+      (Db.PIF.get_IP ~__context ~self:pif) :: (Db.PIF.get_IPv6 ~__context ~self:pif)
```

Comment:
Consider omitting the first item from this list if `get_IP` returns the empty string.
(Equivalently, pipe the whole lot of `ips` through a filter to remove any empty strings.)